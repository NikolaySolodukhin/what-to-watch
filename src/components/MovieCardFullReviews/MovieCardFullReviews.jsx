import React, {PureComponent} from 'react';
import PropTypes from 'prop-types';

const options = {year: `numeric`, month: `long`, day: `numeric`};

class MovieCardFullReviews extends PureComponent {
  static renderItem(array) {
    return array.map((item) => (
      <div className="review" key={item.comment} >
        <blockquote className="review__quote">
          <p className="review__text">{item.comment}</p>

          <footer className="review__details">
            <cite className="review__author">{item.user.name}</cite>
            <time className="review__date">{MovieCardFullReviews.transformDate(item.date)}</time>
          </footer>
        </blockquote>

        <div className="review__rating">{item.rating}</div>
      </div>));
  }

  static transformDate(date) {
    const dateUTC = new Date(date);
    return new Intl.DateTimeFormat(`en-US`, options).format(dateUTC);
  }

  render() {
    const {reviews} = this.props;
    const evenReviewsList = reviews.filter((element, index) => index % 2 === 0);
    const oddReviewsList = reviews.filter((element, index) => index % 2 !== 0);
    return (
      <div className="movie-card__reviews movie-card__row">
        <div className="movie-card__reviews-col">{MovieCardFullReviews.renderItem(evenReviewsList)}</div>
        <div className="movie-card__reviews-col">{MovieCardFullReviews.renderItem(oddReviewsList)}</div>
      </div>
    );
  }
}

MovieCardFullReviews.propTypes = {
  film: PropTypes.shape({
    id: PropTypes.number.isRequired,
    backgroundColor: PropTypes.string.isRequired,
    backgroundImage: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    director: PropTypes.string.isRequired,
    genre: PropTypes.string.isRequired,
    isFavorite: PropTypes.bool.isRequired,
    name: PropTypes.string.isRequired,
    posterImage: PropTypes.string.isRequired,
    previewImage: PropTypes.string.isRequired,
    previewVideoLink: PropTypes.string.isRequired,
    rating: PropTypes.number.isRequired,
    released: PropTypes.number.isRequired,
    runTime: PropTypes.number.isRequired,
    scoresCount: PropTypes.number.isRequired,
    starring: PropTypes.array.isRequired,
    videoLink: PropTypes.string.isRequired
  }),
  reviews: PropTypes.arrayOf(
      PropTypes.shape({
        comment: PropTypes.string.isRequired,
        date: PropTypes.string.isRequired,
        id: PropTypes.number.isRequired,
        rating: PropTypes.number.isRequired,
        user: PropTypes.shape({
          id: PropTypes.number.isRequired,
          name: PropTypes.string.isRequired
        }).isRequired,
      }).isRequired
  ).isRequired,
};

export default MovieCardFullReviews;
