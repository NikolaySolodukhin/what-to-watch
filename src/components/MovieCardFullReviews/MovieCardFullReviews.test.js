import React from 'react';
import Enzyme, {shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import {shallowToJson} from 'enzyme-to-json';

import MovieCardFullReviews from './MovieCardFullReviews';

Enzyme.configure({adapter: new Adapter()});

describe(`<MovieCardFullReviews/>`, () => {
  it(`MovieCardFullReviews renders correctly`, () => {
    const mockProps = {
      film: {
        id: 1,
        genreId: 1,
        backgroundColor: `white`,
        backgroundImage: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
        description: `Film`,
        director: `Raccoon`,
        genre: `comedian`,
        isFavorite: false,
        name: `Fantastic Beasts`,
        posterImage: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
        previewImage: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
        previewVideoLink: `img/fantastic-beasts-the-crimes-of-grindelwald.jpg`,
        rating: 4,
        released: 2018,
        runTime: 1990,
        scoresCount: 10,
        starring: [],
        videoLink: `https://download.blender.org/durian/trailer/sintel_trailer-480p.mp4`
      },
      reviews: [{
        comment: `Unfortunately we don't have a reliable way to tell the true ratings of a movie.`,
        date: `2019-06-18T20:42:47.046Z`,
        id: 1,
        rating: 2,
        user: {
          id: 10,
          name: `Max`
        }
      }]};
    const tree = shallow(<MovieCardFullReviews film={mockProps.film} reviews={mockProps.reviews} />);
    expect(shallowToJson(tree)).toMatchSnapshot();
  });
});
