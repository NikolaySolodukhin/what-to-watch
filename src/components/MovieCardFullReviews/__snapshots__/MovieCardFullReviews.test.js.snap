// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MovieCardFullReviews/> MovieCardFullReviews renders correctly 1`] = `
<div
  className="movie-card__reviews movie-card__row"
>
  <div
    className="movie-card__reviews-col"
  >
    <div
      className="review"
      key="Unfortunately we don't have a reliable way to tell the true ratings of a movie."
    >
      <blockquote
        className="review__quote"
      >
        <p
          className="review__text"
        >
          Unfortunately we don't have a reliable way to tell the true ratings of a movie.
        </p>
        <footer
          className="review__details"
        >
          <cite
            className="review__author"
          >
            Max
          </cite>
          <time
            className="review__date"
          >
            June 18, 2019
          </time>
        </footer>
      </blockquote>
      <div
        className="review__rating"
      >
        2
      </div>
    </div>
  </div>
  <div
    className="movie-card__reviews-col"
  />
</div>
`;
